### PR TITLE
sway: 1.2 -> 1.4, wlroots: 0.8.1 -> 0.10.0

### PR DESCRIPTION
**Pull request by primeos**
###### Motivation for this change

Ready for final testing :)
On my setup I've successfully tested `sway`, `swayidle`, `swaylock`, and `cage`.

`nix-review`:
```
8 package were build:
cage sway sway-unwrapped swayidle swaylock swaylock-fancy waybar wlroots
```

TODOs:
- [x] Testing (also e.g. waybar and cage)
- [x] Squash the RC commits before merging
- [x] The default terminal was replaced with Alacritty (needs to be reflected in the module, we probably have to keep urxvt for backwards compatibility)
- [x] Check compatibility with wayfire (https://github.com/NixOS/nixpkgs/pull/76759)
  - **Doesn't build with wlroots 0.9.1 yet**
- [x] Fix the cage build / request a new upstream release:
  ```
  builder for '/nix/store/2az706idl7hd63yfhlvr66vsn6q75yp1-cage-0.1.1.drv' failed with exit code 1; last 10 log lines:
      wlr_data_device_manager_destroy(data_device_mgr);
      ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
      wlr_data_device_manager_create
    ../cage.c:385:2: error: implicit declaration of function 'wlr_compositor_destroy'; did you mean 'wlr_cursor_destroy'? [-Werror=implicit-function-declaration]
      wlr_compositor_destroy(compositor);
      ^~~~~~~~~~~~~~~~~~~~~~
      wlr_cursor_destroy
    cc1: all warnings being treated as errors
    [4/9] Compiling C object 'cage@exe/output.c.o'.
    ninja: build stopped: subcommand failed.
  ```
  WIP: `master` is fine and I've requested a new release: https://github.com/Hjdskes/cage/issues/107
  -> Will use master for now.

Will be done late in a separate PR:
- [ ] [Grimshot](https://github.com/swaywm/sway/blob/8ffa3cf43906fa95c127c01b751590e77c7bf695/contrib/grimshot) (should we ship the dependencies by default?)
  - Extra tools in `contrib/` aren't installed by default but we could e.g. create an additional package like `sway-contrib`. (or sway-tools, sway-extras, etc.).
###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [x] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

###### Notify maintainers

cc @

### CODE REVIEW

**Review by Synthetica9 — APPROVED** [review]
Works on my machine :+1:

**Comment by Elyhaka** [comment]
Hello @primeos : I have tested the PR, and it seems to work properly as of `1.3-rc1`. It works correctly (and it fixed some issues I had with sway previously).

I can confirm that `waybar` is working properly.

**Comment by Elyhaka** [comment]
@primeos : Should we add `grimshot` like we've added `swaybg` in the wrapper ? Maybe add a flag specifically for enabling it ? Maybe we could add a flag also for `swaybg`, wdyt ?

**Comment by Ma27** [comment]
> Should we add grimshot like we've added swaybg in the wrapper ? Maybe add a flag specifically for enabling it ? Maybe we could add a flag also for swaybg, wdyt ?

I think we should add those by default for convenience. I originally added `swaybg` to the module after I was fairly confused why `swaybg` didn't work. In case of a too bloated closure size we could add e.g. a `minimal` flag (or a `sway-minimal`).

**Review by Ma27 — APPROVED** [review]

**Comment by Synthetica9** [comment]
Now crashes with:

```
2020-01-09 18:01:29 - [EGL] command: eglInitialize, error: 0x3001, message: "DRI2: failed to add configs"
2020-01-09 18:01:29 - [EGL] command: eglInitialize, error: 0x3001, message: "eglInitialize"
2020-01-09 18:01:29 - [render/egl.c:190] Failed to initialize EGL
2020-01-09 18:01:29 - [EGL] command: eglMakeCurrent, error: 0x3008, message: "Invalid display (nil)"
2020-01-09 18:01:29 - [render/wlr_renderer.c:221] Could not initialize EGL
2020-01-09 18:01:29 - [backend/drm/renderer.c:40] Failed to create EGL/WLR renderer
2020-01-09 18:01:29 - [backend/drm/backend.c:203] Failed to initialize renderer
2020-01-09 18:01:29 - [backend/backend.c:198] Failed to open DRM device 9
2020-01-09 18:01:29 - [backend/backend.c:343] Failed to open any DRM device
2020-01-09 18:01:29 - [sway/server.c:47] Unable to create backend
```

For both 1.3-rc1 and 1.3-rc2. 1.2 works fine.

**Comment by primeos** [comment]
@Synthetica9 @Elyhaka @Ma27 huge thanks for testing this PR and providing feedback! :heart:

@Synthetica9 strange that both Sway release candidates do now crash :o Was that due to the second wlroots update from 0.9.0 to 0.9.1? On my setup Sway 1.3-rc2 still works fine.

**Comment by Synthetica9** [comment]
> @Synthetica9 strange that both Sway release candidates do now crash :o Was that due to the second wlroots update from 0.9.0 to 0.9.1? On my setup Sway 1.3-rc2 still works fine.

I think I may have incorrectly tested 1.3-rc1, running 1.2 when I thought I was running 1.3-rc1.

**Comment by Synthetica9** [comment]
Filed an upstream issue: https://github.com/swaywm/sway/issues/4898

**Comment by Elyhaka** [comment]
Still no issue on my side :

image

Could your issue be GPU-related @Synthetica9 ?

**Comment by Synthetica9** [comment]
> Could your issue be GPU-related @Synthetica9 ?

I suppose it _could_ be, but I don't suspect it since I have an Intel IGPU laptop. How would I even diagnose that?

**Comment by Synthetica9** [comment]
@primeos Could you rebase on master? To get ad8a430a3ae9d12508b26825ff2286dfd228cbb2

**Comment by primeos** [comment]
@Synthetica9 good idea, forgot about that, thanks :)

**Review by Synthetica9 — APPROVED** [review]
... And now it fixed itself again. Current working hypothesis: PEBCAK (somewhere)

**Comment by primeos** [comment]
(Did another rebase to re-trigger CI as the evaluation was broken.)

> And now it fixed itself again. Current working hypothesis: PEBCAK (somewhere)

Cool, glad it works now :) Maybe there where also some other relevant dependency updates in the meantime that fixed it (like the Mesa update in my case).

File: pkgs/applications/window-managers/cage/default.nix [hunk 1]
```diff
@@ -1,19 +1,25 @@
...
 , wlroots, wayland, wayland-protocols, pixman, libxkbcommon
 , systemd, libGL, libX11
 , xwayland ? null
 }:
 
 stdenv.mkDerivation rec {
-  pname = "cage";
-  version = "0.1.1";
+  pname = "cage-unstable";
+  version = "2020-01-18";
```
Comment (Synthetica9) [inline pkgs/applications/window-managers/cage/default.nix]: I believe the convention is to have the pname stay the same, and change the version to "unstable-yyyy-mm-dd"

Comment (primeos) [reply pkgs/applications/window-managers/cage/default.nix]: That's actually a common misconception, see: https://github.com/NixOS/nixpkgs/issues/68518

But it does seem a bit strange and there are arguments for both variants (though I also tend towards `pname` as version should start with a digit).

Comment (Synthetica9) [reply pkgs/applications/window-managers/cage/default.nix]: Oh, that's kinda counter-intuitive, but sure

**Comment by primeos** [comment]
After some testing/reviews this should be finally ready to be merged :tada:.

Regarding cage: 0.1.1 doesn't build with wlroots 0.10.0 and IMO backporting the patches is dangerous (I'm not familiar with the codebase and might miss other important changes, but e.g. FreeBSD seems to do this successfully [patch-wlroots-0.9](https://github.com/freebsd/freebsd-ports/blob/6f00c075fe2ef9421c6d5be722c7b81c7dc67841/x11-wm/cage/files/patch-wlroots-0.9)). Therefore I decided to use the master version. An alternative would e.g. be to override wlroots back to the old (compatible) version. Any strong opinions regarding this?

**Review by Synthetica9 — APPROVED** [review]
Works for me \*\o/\*

**Comment by Elyhaka** [comment]
Everything's running smoothly here :)

**Comment by primeos** [comment]
Great, thanks everyone for the testing and valuable feedback :)

:tada: Merged now :tada:

**Comment by NilsIrl** [comment]
How can I upgrade mesa? (I try to keep everything related to `nixos-unstable` using `nix-env`)

**Comment by primeos** [comment]
@NilsIrl normally/officially you can only update it system-wide (i.e. update the channel and `nixos-rebuild`). Using Sway from `nixos-unstable` on a system based on the stable channel is not really supported in general (it often works anyway, but might result in such problems/incompatibilities).

You could upgrade Mesa independently via setting `hardware.opengl.package` to an appropriate value (see `nixos/modules/hardware/opengl.nix`) but that could cause other problems. And there are also other hacks available (e.g. `$LD_LIBRARY_PATH`).

**Comment by NilsIrl** [comment]
I tried setting `LD_LIBRARY_PATH` but it didn't work.

and from what I can tell, it was fixed on its own for @Synthetica9 so it might have nothing to do with mesa.

**Comment by Synthetica9** [comment]
I think mine fixed itself after the merge with master, so it might have
been an mismatched mesa version that fixed itself at that point. (I run
nixos-unstable as my system channel)

**Comment by NilsIrl** [comment]
Well, running everything under `nixos-unstable` does fix the problem.

**Comment by primeos** [comment]
FYI: I've opened #87979 regarding the packaging of the tools in `contrib/` (mainly Grimshot).

(Thought I'd also mention this here since we discussed it a bit and it was an optional TODO:)
> Will be done later in a separate PR:
> - [ ] Grimshot (should we ship the dependencies by default?)
>   - Extra tools in contrib/ aren't installed by default but we could e.g. create an additional package >     like sway-contrib. (or sway-tools, sway-extras, etc.).

**Pull request merged**